core/corehttp/gateway_indexPage.go

```diff
@@ -11,18 +11,51 @@ import (
 
 // structs for directory listing
 type listingTemplateData struct {
-	Listing  []directoryItem
-	Path     string
-	BackLink string
-	Hash     string
+	GatewayURL  string
+	Listing     []directoryItem
+	Size        string
+	Path        string
+	Breadcrumbs []breadcrumb
+	BackLink    string
+	Hash        string
 }
 
 type directoryItem struct {
-	Size string
+	Size      string
+	Name      string
+	Path      string
+	Hash      string
+	ShortHash string
+}
+
+type breadcrumb struct {
 	Name string
 	Path string
 }
 
+func breadcrumbs(path string) []breadcrumb {
+	var ret []breadcrumb
+	var pathParts = strings.Split(path, "/")
+	for i, pathPart := range pathParts {
+		if pathPart == "" {
+			continue
+		}
+		if pathPart == "ipfs" {
+			ret = append(ret, breadcrumb{Name: pathPart})
```

Comment:
This `for` loop needs to be refactored. 

I believe we want to exclude first namespace (`/ipfs/` and `/ipns/`) from being turned into a link **only if it is the very first segment**. 

Right now it works only for `/ipfs/` :

> ![2020-07-22--01-01-55](https://user-images.githubusercontent.com/157609/88115682-06b1f100-cbb7-11ea-8cc5-6b20046c7a47.png)
> ![2020-07-22--01-01-42](https://user-images.githubusercontent.com/157609/88115694-0fa2c280-cbb7-11ea-85a3-5bfcf1a76a50.png)

However, the code is too naive, and disables link for subdirectory:

> ![2020-07-22--01-12-48](https://user-images.githubusercontent.com/157609/88116245-783e6f00-cbb8-11ea-8d9d-cb4e5a7f7f51.png)


Test: http://bafybeig2ed6fx4mzkgopqafswqgjclffnscfphwsjmajt2otat5wc2zaim.ipfs.localhost:8080/foo/ipfs/bar

Reply:
Thanks @lidel, you are correct. I have updated the `breadcrumbs` function to correctly handle any valid path and the links you provided should now work as expected.

Reply:
Sweet, breadcrumbs look great now! 

> ![Screenshot_2020-07-23 Files on IPFS(1)](https://user-images.githubusercontent.com/157609/88327464-5fa09700-cd27-11ea-8768-9f5e51027b94.png)


> ![Screenshot_2020-07-23 Files on IPFS](https://user-images.githubusercontent.com/157609/88327293-37189d00-cd27-11ea-97b7-8f6c82e60056.png)